feat(datasets): Refactored the ManagedTableDataset by Separating Out Common Table Logic

MinuraPunchihewa opened the pull request:
## Description
This PR refactors the `ManagedTableDataset` by separating our the common 'table' (managed and external) logic to a separate `BaseTableDataset`.

## Development notes
The common table logic that is used to execute operations on Databricks Unity Catalog has been separated out here, with the primary purpose to be the implementation of an `ExternalTableDataset` in the near future.

## Checklist

- [X] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the relevant `RELEASE.md` file
- [x] Added tests to cover my changes

MinuraPunchihewa commented:
Hey @astrojuanlu,
As I have mentioned in the description while working on adding the `ExternalTableDataset`, due to some of the many similarities of how the operations for both managed and external tables work, I created a `BaseTableDataset` that contains this common logic and then extended both the `ExternalTableDataset` and the existing `ManagedTableDataset` from it. I made a few slight improvements to the existing implementation of the `ManagedTableDataset` as well.

I have tested everything manually it seems to work fine, but I was late in reading the contributing guidelines and I realize that I should have added all of this under `kedro_datasets_experimental`. If you want me to update this PR by placing my code there, please let me know or if it can be reviewed like this, I will work on adding the unit tests and solving the CI issues to complete the PR. 

I apologize for not reading the contribution guidelines beforehand. That was my mistake.

P.S I am more than happy to maintain this dataset moving forward.

MigQ2 commented on kedro-datasets/kedro_datasets/databricks/external_table_dataset.py:
```diff
@@ -0,0 +1,184 @@
+            catalog: The catalog of the table.
+            database: The database of the table.
+            format: The format of the table.
+            write_mode: The write mode for the table.
```
Minor thinge here: I realized `location` is missing from the docstring

noklam reviewed (commented):
Lots of great discussion here @MigQ2 @MinuraPunchihewa. I'd very love to see this goes in as quick as possible. Here are some suggestions to make this PR review process quicker.

For the `ExternalTableDataset`, we will take this as an `experimental` dataset https://github.com/kedro-org/kedro-plugins/issues/629.

This requires some change but we can assist that so don't worry about this. I'd suggest:
- Breakdown the refactoring part into its own PR (_base_table_dataset.py and managed_table_dataset.py), since I expect the discussion will take longer to fully resolve the new external dataset, the rest can be reviewed separately. (The PR is quite big in size already)


Some more note about[ Experimental Dataset](https://github.com/kedro-org/kedro-plugins/blob/main/kedro-datasets/CONTRIBUTING.md#experimental-datasets)

MinuraPunchihewa commented:
Hey @noklam,
Thank you for your review; so if I have understood correctly, what you want me to is as follows, yes?
1. Separate the refactor of the `ManagedTableDataset` using `BaseTableDataset` into its own PR.
2. Create another PR for the `ExternalTableDataset` (as an experimental dataset) in its full form, i.e., without extending the `BaseTableDataset`?
3. In time, hopefully, we can `BaseTableDataset` to extend `ExternalTableDataset` and clean it up.

Is this correct? If so, can this PR please be the one for the refactoring changes? I ask because that is where most of the work here went into. I can create a different one for the `ExternalTableDataset`.

noklam commented:
@MinuraPunchihewa Correct, maybe we just need 1, 2 is optionally. I can get 1 reviewed quickly if you can do it this week, then you can just focus on this PR (it should be mostly change about `external_table_dataset.py` once 1 is merged.

MinuraPunchihewa commented:
@noklam Thank you, I understand. Then let me rename this PR, remove all changes related to the `ExternalTableDataset` implementation. and keep it for 1. I will create another PR later (for `external_table_dataset.py`) once this one is reviewed and merged.

MinuraPunchihewa commented:
@noklam As requested, I have updated this PR to only contain the changes related to the refactor. From the CI runs, it is not entirely clear to me why the lint and unit tests are failing though. Is there anything else that I need to do here?

noklam commented on kedro-datasets/kedro_datasets/databricks/managed_table_dataset.py:
```diff
@@ -236,225 +102,104 @@ def __init__(  # noqa: PLR0913
-            catalog=catalog,
+        super().__init__(
             table=table,
+            catalog=catalog,
+            database=database,
```
Can the order be preserved? this would be a breaking change.

MinuraPunchihewa replied:
I just pushed a commit. Does this resolve your concern? Do the order of operations matter here since I am passing in keyword arguments?

noklam replied:
The dataset class order doesn't matter because it is already a keyword only class.

The manage table class however is technically still a public class so change of order is a breaking change. It's not gonna affect too much since it's rare but I prefer only introduce breaking change when necessary.

MinuraPunchihewa replied:
@noklam Very clear. Thank you.

noklam reviewed (commented):
Generally looks fine for me. If you have to add new argument, make sure it's a non-breaking one, i.e. add it in the end instead of changing the existing order.

ankatiyar commented:
Hey @MinuraPunchihewa, the refactoring looks good to me but the test coverage for `kedro_datasets/databricks/_base_table_dataset.py` is not complete which is causing the CI to fail.

MinuraPunchihewa commented:
@ankatiyar @noklam I am working on the tests now. Can you please let me know how I can run them?
I tried `make plugin=kedro-datasets test` and `pytest tests/databricks/test_managed_table_dataset.py`, but I get the following error:
```
ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: unrecognized arguments: --cov-report --cov-report term-missing --cov kedro_datasets --cov tests --no-cov-on-fail tests --cov-config pyproject.toml --numprocesses 4 --dist loadfile
  inifile: /home/minura/Documents/work/kedro/repos/kedro-plugins/kedro-datasets/pyproject.toml
  rootdir: /home/minura/Documents/work/kedro/repos/kedro-plugins/kedro-datasets
```

Is it possible to run only a particular test module with make?

noklam commented:
Have you run `make plugin=kedro-datasets install-test-requirements` already? You are [missing the testing dependencies](https://github.com/kedro-org/kedro-plugins/blob/312bd9940aca3f2f492269bb43922c4e0671a4aa/kedro-datasets/pyproject.toml#L202) (the error is complainign pytest-cov)

MinuraPunchihewa commented:
@noklam Thank you; I was able to get it running after installing the dependencies.
@ankatiyar  added some tests; what I basically did was take the existing tests for `managed_table_dataset.py` and moved it to `_base_table_dataset.py`. Then I added a simple tests for the describe() method of `managed_table_dataset.py` because that is the only place where it differs. Let me know what you think.
I've fixed the lint issues as well.

MinuraPunchihewa commented:
I don't understand why the tests are failing.

MinuraPunchihewa commented:
Hey @noklam, @ankatiyar,
I needed to add a couple of other tests to get the coverage to 100%, but a couple of these tests require an external location to save the external tables in. I've added this in as a fixture which takes the location from an environment variable called `DATABRICKS_EXTERNAL_LOCATION`. Is it possible to set this up in Databricks environment configured for testing or is there a better way to go about this?

ankatiyar reviewed (approved):
Thanks @MinuraPunchihewa!

ankatiyar commented:
I dont believe we have a test databricks setup but we can look at how to deal with the tests when you're `ExternalTableDataset` PR is ready. I'll try to get this refactor in for now.

MinuraPunchihewa commented:
> I dont believe we have a test databricks setup but we can look at how to deal with the tests when you're `ExternalTableDataset` PR is ready. I'll try to get this refactor in for now.

Thank you, @ankatiyar. Super stoked to have this merged.
Cool. May I know how the unit tests are running though? The unit tests that existed for `ManagedTableDataset` seemed to require a Spark environment to run from what I could gather.
